lib/commands/publish.js

```diff
@@ -1,6 +1,173 @@
 'use strict'
 
+const assert = require('assert').strict
+const pino = require('pino')
+const { validate } = require('../validation')
+const draft = require('./draft')
+const Npm = require('../npm')
+const GitDirectory = require('../git-directory')
+const GitHub = require('../github')
+
+// TODO this could be optimized with a common schema using $ref..
+const ARGS_SCHEMA = {
+  type: 'object',
+  required: ['path', 'verbose', 'major', 'remote', 'branch', 'semver', 'ghToken'],
+  properties: {
+    major: { type: 'boolean' },
+    help: { type: 'boolean' },
+    dryRun: { type: 'boolean' },
+    ghReleaseDraft: { type: 'boolean' },
+    ghReleasePrerelease: { type: 'boolean' },
+    path: { type: 'string' },
+    tag: { type: 'string' },
+    remote: { type: 'string' },
+    branch: { type: 'string' },
+    ghToken: {
+      type: 'string',
+      minLength: 40
+    },
+    verbose: {
+      type: 'string',
+      enum: ['debug', 'info', 'warn', 'error']
+    },
+    npmAccess: {
+      type: 'string',
+      enum: ['public', 'restricted']
+    },
+    npmDistTag: {
+      type: 'string',
+      pattern: '^[^v0-9][\\w]{1,12}'
+    },
+    semver: {
+      type: 'string',
+      enum: ['major', 'premajor', 'minor', 'preminor', 'patch', 'prepatch', 'prerelease']
+    }
+  }
+}
+
+const CLEAN_REPO = {
+  not_added: [],
+  conflicted: [],
+  created: [],
+  deleted: [],
+  modified: [],
+  renamed: [],
+  files: [],
+  staged: [],
+  ahead: 0,
+  behind: 0,
+  tracking: null
+}
+
+// TODO refactor to make optionals step (publish / push / release)
 module.exports = async function (args) {
-  // TODO
-  // [NOTE for PUBLISH command] if major => block it (unless there is a boolean option)
+  validate(args, ARGS_SCHEMA)
+
+  const logger = pino({ level: args.verbose, prettyPrint: true, base: null })
+
+  const git = GitDirectory(args.path)
+  const status = await git.status()
+  try {
+    const compare = { ...status }
+    delete compare.current // the current branch could be named in a different way
+    CLEAN_REPO.tracking = `${args.remote}/${args.branch}` // your local branch must track where you will push
+    assert.deepStrictEqual(compare, CLEAN_REPO, 'The git repo must be clean (committed and pushed) before releasing!')
+  } catch (assert) {
+    logger.debug({ status }, 'Complete repo status')
+    throw assert
+  }
```

Comment:
this try-catch is useless. Move the logger call before the assert.

Reply:
My intent was to print the status only if necessary, I will fix with a better log message and removing the try/catch 👍